timemachine/cpp/src/gpu_utils.cuh

```diff
@@ -83,3 +84,81 @@ double __device__ __forceinline__ rmul_rn(double a, double b) { return __dmul_rn
 float __device__ __forceinline__ radd_rn(float a, float b) { return __fadd_rn(a, b); }
 
 double __device__ __forceinline__ radd_rn(double a, double b) { return __dadd_rn(a, b); }
+
+// If this were not int128s, we could do __shfl_down_sync, but only supports up to 64 values
+__device__ __forceinline__ void warp_reduce(volatile __int128 *shared_data, int thread_idx) {
+#pragma unroll 6
+    for (int i = 32; i > 0; i /= 2) {
+        if (thread_idx < i) {
+            shared_data[thread_idx] = shared_data[thread_idx] + shared_data[thread_idx + i];
+        }
+        __syncwarp();
+    }
+}
+
+// Taken originally from https://developer.download.nvidia.com/assets/cuda/files/reduction.pdf
+// but modified to account for post Cuda 9.0 which no longer implicitly syncs warps and the
+// fact that we don't need to launch recursive blocks as we can simply reduce the number of energies
+// that are accumulated.
+template <unsigned int BLOCK_SIZE>
+void __global__ k_accumulate_energy(
```

Comment:
Can we add a unit test just for this kernel? (Either in C++ or Python)

Reply:
Working on this, but a bit more involved to expose the kernels to the warp_kernels due to where the logic is currently and what is kernel only code (float -> fixed) and such. 

Want to investigate gtest at some point, but would be best in a separate pass

Reply:
I do think we need to test some of this logic in isolation (like how we do with the neighborlist code, even though the NB code uses them):

1) so we can sleep better at night
2) so we can iterate quickly on it when we add support for more specialized shuffles later on

I think we can just add a wrapper function and expose it to `wrap_kernels.cpp` , similar to:

https://github.com/proteneer/timemachine/blob/master/timemachine/cpp/src/wrap_kernels.cpp#L1252

Reply:
Added test in https://github.com/proteneer/timemachine/pull/1090/commits/adf5700639c9631f7e7d865af2782e6732b95266